File: src/main/java/com/battletrident/games/ring/RingManager.java
package com.battletrident.games.ring;

import org.bukkit.WorldBorder;

import static com.battletrident.consts.Plugin.*;

public class RingManager {
	private final WorldBorder worldBorder;
	private final PhaseManager phase;

	public RingManager() {
		worldBorder = getWorld().getWorldBorder();
		worldBorder.setWarningDistance(30);
		worldBorder.setWarningTime(0);
		worldBorder.setDamageBuffer(0);

		phase = new PhaseManager();
	}

	private void setRing(Ring ring) {
		worldBorder.setSize(ring.size(), ring.speed());
		worldBorder.setDamageAmount(ring.damage());
	}
	private void repeat() {
		Ring current = phase.current();
		setRing(current);

		current.ender = runTaskLater(() -> {
			phase.notifyEnd();

			if (phase.hasMore()) {
				current.starter = runTaskLater(() -> {
					phase.next();
					repeat();
				}, current.delay() * 20);
			}
		}, current.speed() * 20);
	}

	public void reset() {
		worldBorder.setCenter(getWorld().getSpawnLocation());

		phase.clearTasks();
		phase.set(0);
		setRing(phase.current());
	}

	public void start() {
		reset();

		repeat();
	}
}

File: src/main/kotlin/com/battletrident/games/ring/RingManager.kt
package com.battletrident.games.ring

import com.battletrident.BattleTrident.Companion.plugin
import com.battletrident.BattleTrident.Companion.world
import org.bukkit.WorldBorder
import org.bukkit.scheduler.BukkitRunnable

class RingManager {
	private val worldBorder: WorldBorder = world.worldBorder
	private val phase: PhaseManager

	init {
		worldBorder.warningDistance = 30
		worldBorder.warningTime = 0
		worldBorder.damageBuffer = 0.0

		phase = PhaseManager()
	}

	private fun setRing(ring: Ring) {
		worldBorder.setSize(ring.size, ring.speed)
		worldBorder.damageAmount = ring.damage
	}

	private fun repeat() {
		val current: Ring = phase.current()
		setRing(current)

		current.ender = object : BukkitRunnable() {
			override fun run() {
				phase.notifyEnd()
				
				if (phase.hasMore()) {
					current.starter = object : BukkitRunnable() {
						override fun run() {
							phase.next()
							repeat()
						}
					}
					
					current.starter?.runTaskLater(plugin, current.delay * 20)
				}
			}
		}
		current.ender?.runTaskLater(
			plugin,
			current.speed * 20
		)
		
		/*
		current.ender = runTaskLater({
			phase.notifyEnd()
			if (phase.hasMore()) {
				current.starter = runTaskLater({
					phase.next()
					repeat()
				}, current.delay * 20)
			}
		}, current.speed * 20)
		 */
	}

	fun reset() {
		worldBorder.center = world.getSpawnLocation()

		phase.clearTasks()
		phase.set(0)
		setRing(phase.current())
	}

	fun start() {
		reset()

		repeat()
	}
}